docs/design/SecretsEncryption.md

```diff
@@ -0,0 +1,192 @@
+    * bucket-master-key - stored in the DB, encrypted by system-master-key, and used to encrypt all the chunks cipher keys of the bucket.
+ * Key rotation policies:
+    * On demand - API request to rotate all keys or a specific scope (root/system/account/bucket).
+    * Master keys expiry policy - set a system wide policy to expire and rotate keys automatically.
+    * Chunk keys expiry policy - set a bucket/system policy to expire the chunk cipher keys which will re-generate cipher keys and re-encrypt the chunks data on the background.
+
+## OUT OF SCOPE
+Integration with external HSM or KMS is out of scope for this delivery. Using a single system-master-key would provide us an easy interface to add it later on since we would need to worry only about our top level master key.
+
+## SOLUTION
+
+### Hierarchy master keys tree
+The following diagram illustrates a simple key hierarchy and its relationship to NooBaa entities data.
+<div id="top" />
+<img src="/docs/design/images/master_keys_diagram.png" />
+In the figure, “Chunk” and “Account” schemas have the data we ultimately want to protect, such as chunk cipher keys, cipher iv, access keys, etc. For these scenarios, symmetric key encryption (where the same key is used both to encrypt and decrypt the data) is the norm, because it’s efficient.
+
+
+The application encrypts the sensitive data using the desired master key, the desired master key itself encrypted by his master key, and so on recursively.
+
+
+When the application wants to decrypt the sensitive data, it fetches the master key and uses it to decrypt the sensitive data.
+
+
+Keys tree is stored encrypted in the DB memory (on disk memory), and stored decrypted at run-time (on process memory) for performance.
+Structure of master key document in master keys collection:
+
+```
+required: [
+       '_id',
+   ],
+   properties: {
+       _id: { objectid: true },
+       description: { type: 'string' },
+       // If missing - encrypted with "global" key
+       // 1. ENV - Mounted key from Kubernetes secret
+       // 2. HSM configuration TBD - Get key from Vault
+       master_key_id: { objectid: true },
+       // cipher used to provide confidentiality - computed on the compressed data
+       cipher_type: { $ref: 'common_api#/definitions/cipher_type' },
+       cipher_key: { binary: true },
+       cipher_iv: { binary: true },
+       enabled: { type: 'boolean' }
+   }
+```
+
+Notice that each cipher_key and cipher_iv is stored encrypted in DB memory and stored decrypted in process runtime memory (present different values depending on memory location).
+These documents are stored in the system store’s cache, and there is no need to create any transactions to the DB in order for getting them (in memory operations).
+
+### Decryption of master keys at runtime
+The master key hierarchy is decrypted at the system store’s initial load and stored in memory for efficient runtime encryption and decryption of sensitive data.
```

Comment:
We talked about it that we don't want every system_store reload to re-decrypt all keys which is wasteful assuming most keys are not used, so we want it lazy.

Some thoughts:
- Lazy means recursive, but with only two levels of master-keys tree in the system store recursion is not worrysome.
- I think it can be a simple SystemStoreData method, say `get_decrypted_master_key(key_id)`, that checks and returns cached value or otherwise recurse to get master's decrypted key +  decrypt the current key + cache it.
- We can cache it on the same object with a `Symbol('decrypted_master_key')`, or just a `_decrypted_master_key`, or even go as far as having an in memory LRUCache that maps between `cipher_key_b64 => decrypted_cipher_key_b64` so that decryption will be reserved for next reload too.